packages/core-ethereum/crates/core-ethereum-rpc/src/client.rs

```diff
@@ -0,0 +1,278 @@
+#[derive(Debug)]
+pub struct JsonRpcProviderClient<Req: HttpPostRequestor + Debug> {
+    id: AtomicU64,
+    url: String,
+    requestor: Req,
+}
+
+impl<Req: HttpPostRequestor + Debug> JsonRpcProviderClient<Req> {
+    /// Creates the client given the `HttpPostRequestor`
+    pub fn new(base_url: &str, requestor: Req) -> Self {
+        Self {
+            id: AtomicU64::new(1),
+            url: base_url.to_owned(),
+            requestor,
+        }
+    }
+}
+
+impl<Req: HttpPostRequestor + Debug + Clone> Clone for JsonRpcProviderClient<Req> {
+    fn clone(&self) -> Self {
+        Self {
+            id: AtomicU64::new(1),
+            url: self.url.clone(),
+            requestor: self.requestor.clone(),
+        }
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<Req: HttpPostRequestor + Debug> JsonRpcClient for JsonRpcProviderClient<Req> {
+    type Error = JsonRpcProviderClientError;
+
+    async fn request<T: Serialize + Send + Sync, R: DeserializeOwned>(
+        &self,
+        method: &str,
+        params: T,
+    ) -> Result<R, Self::Error> {
+        // Create and serialize the Request object
+        let next_id = self.id.fetch_add(1, Ordering::SeqCst);
+        let payload = Request::new(next_id, method, params);
+        let serialized_payload = serde_json::to_string(&payload).map_err(|err| Self::Error::SerdeJson {
+            err,
+            text: "cannot serialize payload".into(),
+        })?;
+
+        // Perform the actual request
+        let body = self.requestor.http_post(self.url.as_ref(), &serialized_payload).await?;
+
+        // First deserialize the Response object
+        let raw = match serde_json::from_str(&body) {
+            Ok(Response::Success { result, .. }) => result.to_owned(),
+            Ok(Response::Error { error, .. }) => return Err(error.into()),
+            Ok(_) => {
+                let err = Self::Error::SerdeJson {
+                    err: serde::de::Error::custom("unexpected notification over HTTP transport"),
+                    text: body,
+                };
+                return Err(err);
+            }
+            Err(err) => return Err(Self::Error::SerdeJson { err, text: body }),
+        };
+
+        // Next, deserialize the data out of the Response object
+        let res = serde_json::from_str(raw.get()).map_err(|err| Self::Error::SerdeJson {
+            err,
+            text: raw.to_string(),
+        })?;
+
+        Ok(res)
+    }
+}
+
+#[derive(Debug)]
+pub struct SimpleJsonRpcRetryPolicy;
+
+impl SimpleJsonRpcRetryPolicy {
+    fn should_retry_on_json_rpc_error(&self, err: &JsonRpcError) -> bool {
+        let JsonRpcError { code, message, .. } = err;
+
+        // Alchemy throws it this way
+        if *code == 429 {
+            return true;
+        }
+
+        // This is an Infura error code for `exceeded project rate limit`
+        if *code == -32005 {
+            return true;
```

Comment:
Retrying when limit has been reached? Sounds like a counterproductive approach.

Reply:
This is the default approach of `ethers`. Should we change it ?